src/main/java/no/nav/familie/inntektsmelding/server/authz/TilgangsstyringFilter.java

```diff
@@ -0,0 +1,269 @@
+        // Her bør det ligge auditlogging om det blir behov for intern bruker tilgang.
+    }
+
+    @Override
+    public void filter(ContainerRequestContext req) {
+        var method = resourceinfo.getResourceMethod();
+
+        if (ENV.isLocal() && method.getAnnotation(UtenAutentisering.class) != null) {
+            LOG.trace("{} er whitelisted", method);
+            return;
+        }
+
+        if (!KontekstHolder.harKontekst()) {
+            throw new TekniskException("FP-234584", "Kontekst må være satt");
+        }
+
+        var annotation = method.getAnnotation(Tilgangsstyring.class);
+        if (annotation == null) {
+            throw new TekniskException("FP-234584", "Ressurs mangler riktig annotering");
+        }
+
+        var kontekst = KontekstHolder.getKontekst();
+        var policy = annotation.policy();
+        var action = annotation.action();
+
+        // her kjøres selve reglene - tidligere kall til PDP/ABAC - mulig å lage et rammeverk rundt dette - men var overkill til å gjøre dette her.
+        var tilgangBesluttning = switch (kontekst.getIdentType()) {
+            case EksternBruker -> vurderBorgerTilgang(policy, action);
+            case Systemressurs -> vurderSystemTilgang(policy, action);
+            case null, default -> ikkeTilgang(String.format("Mangler policy for %s", kontekst.getIdentType()));
+        };
+
+        if (!tilgangBesluttning.fikkTilgang()) {
+            throw new NektetTilgangException("IKKE-TILGANG", tilgangBesluttning.begrunnelse());
+        } else {
+            LOG.info("Tilgang ok.");
+        }
+    }
+
+    private Tilgangsbeslutning vurderSystemTilgang(PolicyType policy, ActionType action) {
+        var kontekst = KontekstHolder.getKontekst();
+        if (!erAzureProvider(kontekst)) {
+            return ikkeTilgang("Kun Azure brukere støttes.");
+        }
+
+        if (PolicyType.PORTAL.equals(policy)) {
+            if (ActionType.WRITE.equals(action)) {
+                var konsumentUid = kontekst.getUid();
+                if (erPreautorisert(konsumentUid) && erISammeKlusterKlasseOgNamespace(konsumentUid)) {
+                    return okTilgang();
+                }
+                SECURE_LOG.warn("Konsumenten {} er ikke autorisert", konsumentUid);
+                return ikkeTilgang("Konsumenten er ikke autorisert.");
+            }
+        }
+        LOG.warn("Mangler policy for {}, action {}", policy, action);
+        return ikkeTilgang("Mangler policy.");
+    }
+
+    private Tilgangsbeslutning vurderBorgerTilgang(PolicyType policy, ActionType action) {
+        var kontekst = KontekstHolder.getKontekst();
+        if (!erTokenXProvider(kontekst)) {
+            return ikkeTilgang("Kun TokenX brukere støttes.");
+        }
+
+        var input = finnTilgangsstyringInput();
+
+        if (PolicyType.INNTEKTSMELDING.equals(policy)) {
+            if (Set.of(ActionType.READ, ActionType.WRITE).contains(action)) {
```

Comment:
Bør antagelig ta en diskusjon på om behov for sperre mot at innlogget AG kan sende IM for seg selv.
Hva gjør Helse?

Reply:
Noe usikker på hva Helse gjør - men bra poeng.

Reply:
Vi lar det være fritt er visst holdningen